// @truffle/resolver/dist/lib/resolver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resolver = void 0;
const debug_1 = __importDefault(require("debug"));
const debug = debug_1.default("resolver");
const contract = require("@truffle/contract");
const expect = require("@truffle/expect");
const provision = require("@truffle/provisioner");
const sources_1 = require("./sources");
const defaultResolverOptions = {
    includeTruffleSources: false,
    translateJsonToSolidity: true,
    resolveVyperMoudles: false,
};
class Resolver {
    constructor(options, resolverOptions = {}) {
        expect.options(options, ["working_directory", "contracts_build_directory", "contracts_directory"]);
        resolverOptions = Object.assign(Object.assign({}, defaultResolverOptions), resolverOptions);
        const { includeTruffleSources, translateJsonToSolidity, resolveVyperModules } = resolverOptions;
        this.options = options;
        this.sources = [
            new sources_1.EthPMv1(options.working_directory),
            new sources_1.NPM(options.working_directory),
            new sources_1.GlobalNPM(),
            new sources_1.FS(options.working_directory, options.contracts_build_directory)
        ];
        if (includeTruffleSources) {
            this.sources.unshift(new sources_1.Truffle(options));
        }
        if (translateJsonToSolidity) {
            this.sources = [].concat(...this.sources.map(source => [new sources_1.ABI(source), source]));
        }
        if (resolveVyperModules) {
            this.sources = [new sources_1.Vyper(this.sources, options.contracts_directory)];
        }
    }
    // This function might be doing too much. If so, too bad (for now).
    require(import_path, search_path) {
        let abstraction;
        this.sources.forEach((source) => {
            const result = source.require(import_path, search_path);
            if (result) {
                abstraction = contract(result);
                provision(abstraction, this.options);
            }
        });
        if (abstraction)
            return abstraction;
        throw new Error("Could not find artifacts for " + import_path + " from any sources");
    }
    resolve(importPath, importedFrom) {
        return __awaiter(this, void 0, void 0, function* () {
            let body = null;
            let filePath = null;
            let source = null;
            for (source of this.sources) {
                ({ body, filePath } = yield source.resolve(importPath, importedFrom));
                if (body !== undefined) {
                    break;
                }
            }
            if (body === undefined) {
                let message = `Could not find ${importPath} from any sources`;
                if (importedFrom) {
                    message += "; imported from " + importedFrom;
                }
                throw new Error(message);
            }
            return {
                body,
                filePath,
                source
            };
        });
    }
}
exports.Resolver = Resolver;
