// abi-to-sol/dist/test/arbitrary.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_fast_check_1 = require("jest-fast-check");
const abiSchema = __importStar(require("@truffle/contract-schema/spec/abi.spec.json"));
const jest_json_schema_1 = require("jest-json-schema");
expect.extend(jest_json_schema_1.matchers);
const Arbitrary = __importStar(require("./arbitrary"));
const withDefinitions = (schema) => (Object.assign({ definitions: abiSchema.definitions }, schema));
const arbitraries = {
    Type: {
        arbitrary: Arbitrary.Type(),
        schema: abiSchema.definitions.Type,
    },
    Parameter: {
        arbitrary: Arbitrary.Parameter(),
        schema: withDefinitions(abiSchema.definitions.Parameter),
    },
    EventParameter: {
        arbitrary: Arbitrary.EventParameter(),
        schema: withDefinitions(abiSchema.definitions.EventParameter),
    },
    Event: {
        arbitrary: Arbitrary.Event(),
        schema: withDefinitions(abiSchema.definitions.Event),
    },
    NormalFunction: {
        arbitrary: Arbitrary.NormalFunction(),
        schema: withDefinitions(abiSchema.definitions.NormalFunction),
    },
    ConstructorFunction: {
        arbitrary: Arbitrary.ConstructorFunction(),
        schema: withDefinitions(abiSchema.definitions.ConstructorFunction),
    },
    ReceiveFunction: {
        arbitrary: Arbitrary.ReceiveFunction(),
        schema: withDefinitions(abiSchema.definitions.ReceiveFunction),
    },
    FallbackFunction: {
        arbitrary: Arbitrary.FallbackFunction(),
        schema: withDefinitions(abiSchema.definitions.FallbackFunction),
    },
    Abi: {
        arbitrary: Arbitrary.Abi(),
        schema: abiSchema,
    },
};
for (const [name, { arbitrary, schema }] of Object.entries(arbitraries)) {
    describe(`Arbitrary.${name}`, () => {
        jest_fast_check_1.testProp("validates schema", [arbitrary], (value) => {
            expect(value).toMatchSchema(schema);
        });
    });
}
