// @nodefactory/filsnap-adapter/build/methods.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateGasForMessage = exports.getMessages = exports.sendMessage = exports.signMessageRaw = exports.signMessage = exports.configure = exports.exportPrivateKey = exports.getBalance = exports.getPublicKey = exports.getAddress = void 0;
function sendSnapMethod(request, snapId) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield window.ethereum.send({
            method: snapId,
            params: [
                request
            ]
        });
    });
}
function getAddress() {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "getAddress" }, this.snapId);
    });
}
exports.getAddress = getAddress;
function getPublicKey() {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "getPublicKey" }, this.snapId);
    });
}
exports.getPublicKey = getPublicKey;
function getBalance(denomination) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "getBalance", params: { denomination: denomination } }, this.snapId);
    });
}
exports.getBalance = getBalance;
function exportPrivateKey() {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "exportPrivateKey" }, this.snapId);
    });
}
exports.exportPrivateKey = exportPrivateKey;
function configure(configuration) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "configure", params: { configuration: configuration } }, this.snapId);
    });
}
exports.configure = configure;
function signMessage(message) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "signMessage", params: { message: message } }, this.snapId);
    });
}
exports.signMessage = signMessage;
function signMessageRaw(rawMessage) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "signMessageRaw", params: { message: rawMessage } }, this.snapId);
    });
}
exports.signMessageRaw = signMessageRaw;
function sendMessage(signedMessage) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "sendMessage", params: { signedMessage: signedMessage } }, this.snapId);
    });
}
exports.sendMessage = sendMessage;
function getMessages() {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "getMessages" }, this.snapId);
    });
}
exports.getMessages = getMessages;
function calculateGasForMessage(message) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield sendSnapMethod({ method: "getGasForMessage", params: { message: message } }, this.snapId);
    });
}
exports.calculateGasForMessage = calculateGasForMessage;
