// @nodefactory/filsnap-adapter/build/snap.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetamaskFilecoinSnap = void 0;
const methods_1 = require("./methods");
class MetamaskFilecoinSnap {
    constructor(pluginOrigin) {
        this.getFilecoinSnapApi = () => __awaiter(this, void 0, void 0, function* () {
            return {
                calculateGasForMessage: methods_1.calculateGasForMessage.bind(this),
                configure: methods_1.configure.bind(this),
                exportPrivateKey: methods_1.exportPrivateKey.bind(this),
                getAddress: methods_1.getAddress.bind(this),
                getBalance: methods_1.getBalance.bind(this),
                getMessages: methods_1.getMessages.bind(this),
                getPublicKey: methods_1.getPublicKey.bind(this),
                sendMessage: methods_1.sendMessage.bind(this),
                signMessage: methods_1.signMessage.bind(this),
                signMessageRaw: methods_1.signMessageRaw.bind(this),
            };
        });
        this.pluginOrigin = pluginOrigin;
        this.snapId = `wallet_plugin_${this.pluginOrigin}`;
    }
}
exports.MetamaskFilecoinSnap = MetamaskFilecoinSnap;
