// @textile/hub-threads-client/dist/cjs/index.js
"use strict";
/**
 * A web-gRPC wrapper client for communicating with the web-gRPC enabled Threads ove the Hub APIs.
 * @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var threads_client_1 = require("@textile/threads-client");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return threads_client_1.Action; } });
Object.defineProperty(exports, "Query", { enumerable: true, get: function () { return threads_client_1.Query; } });
Object.defineProperty(exports, "ReadTransaction", { enumerable: true, get: function () { return threads_client_1.ReadTransaction; } });
Object.defineProperty(exports, "Where", { enumerable: true, get: function () { return threads_client_1.Where; } });
Object.defineProperty(exports, "WriteTransaction", { enumerable: true, get: function () { return threads_client_1.WriteTransaction; } });
var client_1 = require("./client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
